pyprep/find_noisy_channels.py

```diff
@@ -212,209 +234,229 @@ def find_bad_by_nan_flat(self):
+        frac_bad : float, optional
+            The minimum proportion of bad windows for a channel to be considered
+            "bad-by-correlation" or "bad-by-dropout". Defaults to ``0.01`` (1% of
+            all windows).
 
         """
-        self.find_bad_by_hfnoise()  # since filtering is performed there
-        usable_idx = self.usable_idx
-        correlation_frames = correlation_secs * self.sample_rate
-        correlation_window = np.arange(correlation_frames)
-        correlation_offsets = np.arange(
-            1, (self.n_samples - correlation_frames), correlation_frames
-        )
-        w_correlation = len(correlation_offsets)
-        channel_correlation = np.ones((w_correlation, self.n_chans_original))
-        noiselevels = np.zeros((w_correlation, self.n_chans_original))
-        channel_deviations = np.zeros((w_correlation, self.n_chans_original))
-        drop = np.zeros((w_correlation, self.n_chans_original))
-        len_correlation_window = len(correlation_window)
-        EEGData = np.transpose(self.EEGData)
-        EEG_new_win = np.reshape(
-            np.transpose(EEGData[0 : len_correlation_window * w_correlation, :]),
-            (self.n_chans_new, len_correlation_window, w_correlation),
-            order="F",
-        )
-        data_win = np.reshape(
-            np.transpose(
-                self.EEGData_beforeFilt[0 : len_correlation_window * w_correlation, :]
-            ),
-            (self.n_chans_new, len_correlation_window, w_correlation),
-            order="F",
-        )
-        for k in range(0, w_correlation):
-            eeg_portion = np.transpose(np.squeeze(EEG_new_win[:, :, k]))
-            data_portion = np.transpose(np.squeeze(data_win[:, :, k]))
-            with np.errstate(invalid="ignore"):  # suppress divide-by-zero warnings
-                window_correlation = np.corrcoef(np.transpose(eeg_portion))
-            abs_corr = np.abs(
-                np.subtract(window_correlation, np.diag(np.diag(window_correlation)))
-            )
-            channel_correlation[k, usable_idx] = _mat_quantile(abs_corr, 0.98, axis=0)
-            with np.errstate(invalid="ignore"):  # suppress divide-by-zero warnings
-                noiselevels[k, usable_idx] = np.divide(
-                    robust.mad(np.subtract(data_portion, eeg_portion), c=1),
-                    robust.mad(eeg_portion, c=1),
-                )
-            channel_deviations[k, usable_idx] = 0.7413 * _mat_iqr(data_portion, axis=0)
-        for i in range(0, w_correlation):
-            for j in range(0, self.n_chans_original):
-                drop[i, j] = np.int(
-                    np.isnan(channel_correlation[i, j]) or np.isnan(noiselevels[i, j])
-                )
-                if drop[i, j] == 1:
-                    channel_correlation[i, j] = 0
-                    noiselevels[i, j] = 0
+        IQR_TO_SD = 0.7413  # Scales units of IQR to units of SD, assuming normality
```

Comment:
if you added a comment on this above, I think it's make sense to repeat it here once more.

An alternative would be a new module `config.py` or `defaults.py`, or `constants.py` or similar, where these things get defined and documented once, and then imported from to other modules. Up to you what you like better for this PR.

Reply:
Yeah, I was conflicted about what to do for these: on the one hand, having the constants defined close to where they're used makes it much easier for someone to read the code and refer back to its definition at a glance. On the other hand, in the case of `IQR_TO_SD` it leads to some redundancy and leaves less room for documentation.

Since I think `IQR_TO_SD` is the only constant that's used in more than one function I'd tend towards keeping them with the relevant code for now, but if we end up having more constants like that in the future a separate module for them would be a good idea.

Reply:
agreed!